### PR TITLE
4.0 upgrade

### PR DESCRIPTION
**Pull request by wilr**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -3,21 +3,29 @@
...
+		"email": "damian@silverstripe.com"
+	}],
 	"require": {
-		"silverstripe/comments": "^2.0"
+		"silverstripe/comments": "^3.0"
 	},
 	"require-dev": {
-		"phpunit/PHPUnit": "~3.7@stable"
-	},
+        "phpunit/PHPUnit": "^5.7"
```
Comment (robbieaverill) [inline composer.json]: Add `squizlabs/php_codesniffer: ^3.0`

File: composer.json [hunk 2]
```diff
@@ -3,21 +3,29 @@
...
+    },
+    "scripts": {
+        "lint": "phpcs src/ tests/",
+        "fix": "php-cs-fixer fix src/ && php-cs-fixer fix tests/"
+    },
```
Comment (robbieaverill) [inline composer.json:20]: we probably don't need to add these, or if we do they should reference a local dependency

File: tests/CommentNotifiableTest.php [hunk 3]
```diff
@@ -1,98 +1,70 @@
...
 
-    protected $extraDataObjects = array(
-        'CommentNotifiableTest_DataObject'
-    );
-    
+    protected static $extra_dataobjects = [
+        CommentNotifiableTestDataObject::class
+    ];
+
     public function setUp()
```
Comment (robbieaverill) [inline tests/CommentNotifiableTest.php]: Let's make this protected now, tearDown too

File: .travis.yml [hunk 4]
```diff
@@ -1,29 +1,33 @@
...
-    - php: 5.4
-      env: DB=MYSQL CORE_RELEASE=3
+      env: DB=MYSQL PHPCS_TEST=1 PHPUNIT_TEST=1
+    - php: 7.0
+      env: DB=PGSQL PHPUNIT_TEST=1
     - php: 7.1
-      env: DB=MYSQL CORE_RELEASE=3.6
+      env: DB=MYSQL PHPUNIT_COVERAGE_TEST=1
+    - php: 7.0
+      env: DB=MYSQL
```
Comment (robbieaverill) [inline .travis.yml:13]: This won't do anything, remove

File: .travis.yml [hunk 5]
```diff
@@ -1,29 +1,33 @@
...
+  - phpenv rehash
+  - phpenv config-rm xdebug.ini
+  - echo 'memory_limit = 2G' >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
+  - composer validate
+  - composer require silverstripe/recipe-cms 1.0.x-dev --no-update
+  - if [[ $DB == PGSQL ]]; then composer require --no-update silverstripe/postgresql:2.0.x-dev; fi
+  - composer install --prefer-dist --no-interaction --no-progress --no-suggest --optimize-autoloader --verbose --profile
+
+  # Bootstrap cms / mysite folder
+  - php ./cms/tests/bootstrap/mysite.php
```
Comment (robbieaverill) [inline .travis.yml]: I don't _think_ we need this, since it's in phpunit.xml.dist

**Review by robbieaverill — CHANGES_REQUESTED** [review]
Nice! Minor comments. Can you also add a codecov.yml

File: templates/SilverStripe/CommentNotifications/CommentEmail.ss [hunk 6]
```diff
@@ -19,7 +19,7 @@
 	<% end_if %>
 </ul>
 
-<blockquote>{$Comment.Comment}</blockquote>
+<blockquote><% with Comment %>$Comment<% end_with %></blockquote>
```
Comment (wilr) [inline templates/SilverStripe/CommentNotifications/CommentEmail.ss]: `$Comment.Comment` seemed to flip SSViewer out..

File: composer.json (hunk 2 above)
Comment (wilr) [reply composer.json:20]: So that's just what comments has so it's consistent.

**Comment by robbieaverill** [comment]
Replaced by #20

**Pull request merged**